### ISSUE TITLE
XYChart: render labels next to points

### ISSUE THREAD

**Issue by leeoniya** [grafana/grafana#57222]
ref: https://github.com/grafana/grafana/issues/54192#issuecomment-1282357108

we should do this because it's a common thing in bubble charts (which can be made using this panel).

a basic version of this is relatively easy when we don't have high data density, but with a few hundred points, we need logic to prioritize what to show, and also try to avoid label collisions by omission or by more complex geometric placement algo.

will need to add another field mapping config to select which field to use for label text.

**Comment by codeincarnate** [comment]
I'm guessing here we should use the same code as is used in the BarChart to detect label overlaps. Maybe time to make that code somewhat more generic?

**Comment by leeoniya** [comment]
i think that will be too slow here. bar charts don't have 20k labels and the labels in barchart that we check for overlap are limited the same bar stack.

we'll probably need to come up with a strategy to select which labels to render up front using other heuristics before checking overlap.

**Comment by codeincarnate** [comment]
That's true, I'd be interested to see if using an [R-tree](https://github.com/mourner/rbush) might provide enough performance since it should theoretically group labels by area. Could be worth invetigating.

**Comment by codeincarnate** [comment]
That's very cool! 😎 I'll need to think about this a bit more. Definitely a lot of different strategies to consider too when showing labels.

**Comment by codeincarnate** [comment]
Using the approach outlined here could work https://github.com/furstenheim/automatic-scatter-labelling

**Comment by leeoniya** [comment]
the animated demo there isnt what i would want to see, honestly 😅. the labels are way too dense and male it really difficult to actually see the poonts, even though they technically dont occlude them. labeling every point might be okay for bubble charts with a couple dozen large-ish points, but with a multi-hundred scatter i'd be mostly interested in the overall density distribution and maybe labeled outliers.

**Comment by codeincarnate** [comment]
It has trade-offs for sure. I think the demo is OK but would definitely suggest looking at the linked paper as it presents an interesting approach. I think there are going to be different scenarios where we'd want to use different strategies potentially.

Presumably for lower point numbers we'd want to show as many labels as possible. I agree with you that for high density plots we're going to want to prioritize distribution and try to highlight (for lack of a better way to put it) "interesting" data points rather than attempting to label everything.

**Comment by leeoniya** [comment]
a major issue with the approach (if the animation is representative) even with relatively low density datasets is that the labels are not close enough to the points they represent. plus you get a bunch of visual clutter of leader lines to deal with.

e.g. the Netherlands label is much closer to Greece than the Greece label, which is super far. i'll look through the paper, but a picture is worth 1k words:

![Screenshot_20230119-020232265 (1)](https://user-images.githubusercontent.com/43234/213387580-90ef8a4b-f88a-4dfd-a16a-ad016cc853fd.jpg)

vs something more sane:

![FIG-SCATTER-LABEL-VJUST-2](https://user-images.githubusercontent.com/43234/213388848-bb54cd76-f183-4902-a458-3532972ef943.png)
